signingscript/task.py

```diff
@@ -142,203 +114,55 @@ def get_suitable_signing_servers(signing_servers, cert_type, signing_formats):
-        tar_cmd = ['tar', 'czvf', abs_to, '.']
-        await utils._execute_subprocess(tar_cmd, cwd=app_dir)
-
-    return to
-
-
-# _extract_zipfile {{{1
-async def _extract_zipfile(context, from_, tmp_dir=None):
-    work_dir = context.config['work_dir']
-    tmp_dir = tmp_dir or os.path.join(work_dir, "unzipped")
-    try:
-        files = []
-        rm(tmp_dir)
-        utils.mkdir(tmp_dir)
-        with zipfile.ZipFile(from_, mode='r') as z:
-            for name in z.namelist():
-                files.append(os.path.join(tmp_dir, name))
-            z.extractall(path=tmp_dir)
-        return files
-    except Exception as e:
-        raise SigningScriptError(e)
-
-
-# _create_zipfile {{{1
-async def _create_zipfile(context, to, files, tmp_dir=None):
-    work_dir = context.config['work_dir']
-    tmp_dir = tmp_dir or os.path.join(work_dir, "unzipped")
-    try:
-        with zipfile.ZipFile(to, mode='w', compression=zipfile.ZIP_DEFLATED) as z:
-            for f in files:
-                relpath = os.path.relpath(f, tmp_dir)
-                z.write(f, arcname=relpath)
-        return to
-    except Exception as e:
-        raise SigningScriptError(e)
-
-
-# detached_sigfiles {{{1
-def detached_sigfiles(filepath, signing_formats):
-    """Get a list of detached sigfile paths, if any, given a file path and signing formats.
-
-    This will generally be an empty list unless we're gpg signing, in which case
-    we'll have detached gpg signatures.
+    output = path
+    # Loop through the formats and sign one by one.
+    for fmt in signing_formats:
+        func = FORMAT_TO_SIGNING_FUNCTION.get(
+            fmt, FORMAT_TO_SIGNING_FUNCTION['default']
+        )
+        log.info("Signing {} with {}...".format(output, fmt))
+        output = await func(context, output, fmt)
+    # We want to return a list
+    if not isinstance(output, (tuple, list)):
+        output = [output]
+    # Log the shas of each output file
+    log_shas(context, output)
+    return output
+
+
+# _sort_formats {{{1
+def _sort_formats(formats):
+    """Order the signing formats.
+
+    Certain formats need to happen before or after others, e.g. gpg after
+    any format that modifies the binary.
 
     Args:
-        filepath (str): the path of the file to sign
-        signing_formats (str): the signing formats the file will be signed with
+        formats (list): the formats to order.
 
     Returns:
-        list: the list of paths of any detached signatures.
+        list: the ordered formats.
 
     """
-    detached_signatures = []
-    for sig_type, sig_ext, sig_mime in utils.get_detached_signatures(signing_formats):
-        detached_filepath = "{filepath}{ext}".format(filepath=filepath,
-                                                     ext=sig_ext)
-        detached_signatures.append(detached_filepath)
-    return detached_signatures
+    for fmt in ("widevine", "widevine_blessed", "gpg"):
```

Comment:
Do you think we should add a comment to specify widevine* must happen before gpg?